pkg/sql/set_zone_config.go

```diff
@@ -962,6 +976,69 @@ func validateZoneAttrsAndLocalities(
 	return nil
 }
 
+// validateZoneLocalitiesForSecondaryTenants performs all the constraint/lease
+// preferences validation for secondary tenants. Secondary tenants are only
+// allowed to reference locality attributes as they only have access to region
+// information via the RegionServer. Even then, they're only allowed to
+// reference the "region" and "zone" tiers.
+//
+// Unlike the system tenant, we also catch typos in and validate prohibited
+// constraints. This is because secondary tenant must operate in the narrow
+// view exposed via the RegionServer and are not allowed to configure arbitrary
+// constraints (required or otherwise).
+func validateZoneLocalitiesForSecondaryTenants(
+	ctx context.Context, getRegions regionsGetter, zone *zonepb.ZoneConfig,
+) error {
+	toValidate := accumulateUniqueConstraints(zone)
+	resp, err := getRegions(ctx, &serverpb.RegionsRequest{})
+	if err != nil {
+		return err
+	}
+	regions := make(map[string]struct{})
+	zones := make(map[string]struct{})
+	for regionName, regionMeta := range resp.Regions {
+		regions[regionName] = struct{}{}
+		for _, zone := range regionMeta.Zones {
+			zones[zone] = struct{}{}
+		}
+	}
+
+	for _, constraint := range toValidate {
+		switch constraint.Key {
+		case "zone":
+			_, found := zones[constraint.Value]
+			if !found {
+				return pgerror.Newf(
+					pgcode.CheckViolation,
+					"zone %q matches no existing nodes within the cluster",
+					constraint.Value,
+				)
+			}
+		case "region":
+			_, found := regions[constraint.Value]
+			if !found {
+				return pgerror.Newf(
+					pgcode.CheckViolation,
+					"region %q matches no existing nodes within the cluster",
```

Comment:
Here and above, instead of mentioning "nodes" (which is not exposed to serverless users), just say that "region %q not found". Speaking off, is there a SHOW ALL REGIONS/ZONES or something tenants would be able to use to figure out what they _can_ use?